File: apps/cms/src/lib/aiProvider.ts
/**
 * AI provider abstraction for the CMS agents (Page Agent, Content Agent).
 * Same idea as apps/api/core/ai_provider.py: one place that knows how to force
 * structured JSON output from either Anthropic or Gemini, picked at runtime via
 * AI_PROVIDER — callers pass a JSON-schema-style object and get a plain object back.
 */
import Anthropic from "@anthropic-ai/sdk";
import { GoogleGenAI, Type } from "@google/genai";

const ANTHROPIC_MODEL = "claude-sonnet-4-5";
const GEMINI_MODEL = "gemini-2.5-flash";

export class AIProviderError extends Error {}

export type AgentChatMessage = { role: "user" | "assistant"; content: string };

export type StructuredCallResult = {
  reply: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  proposal: Record<string, any> | null;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type JsonSchema = Record<string, any>;

function provider(): string {
  return (process.env.AI_PROVIDER || "gemini").toLowerCase();
}

/**
 * Runs one turn of a chat-with-tool-proposal agent: free-text reply, plus an
 * optional structured proposal object when the model calls the update tool.
 */
export async function runAgentTurn(
  systemPrompt: string,
  history: AgentChatMessage[],
  stateContext: string,
  toolName: string,
  toolDescription: string,
  inputSchema: JsonSchema,
): Promise<StructuredCallResult> {
  if (provider() === "gemini") {
    return runGeminiTurn(systemPrompt, history, stateContext, toolName, toolDescription, inputSchema);
  }
  return runAnthropicTurn(systemPrompt, history, stateContext, toolName, toolDescription, inputSchema);
}

async function runAnthropicTurn(
  systemPrompt: string,
  history: AgentChatMessage[],
  stateContext: string,
  toolName: string,
  toolDescription: string,
  inputSchema: JsonSchema,
): Promise<StructuredCallResult> {
  const apiKey = process.env.ANTHROPIC_API_KEY;
  if (!apiKey) {
    throw new AIProviderError("ANTHROPIC_API_KEY is not configured");
  }
  const client = new Anthropic({ apiKey });

  const tool: Anthropic.Tool = {
    name: toolName,
    description: toolDescription,
    input_schema: { ...inputSchema, type: "object" },
  };
  const messages: Anthropic.MessageParam[] = [
    { role: "user", content: stateContext },
    ...history.map((m) => ({ role: m.role, content: m.content }) as Anthropic.MessageParam),
  ];

  const response = await client.messages.create({
    model: ANTHROPIC_MODEL,
    max_tokens: 4096,
    system: systemPrompt,
    tools: [tool],
    messages,
  });

  let reply = "";
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let proposal: Record<string, any> | null = null;
  for (const block of response.content) {
    if (block.type === "text") {
      reply += block.text;
    } else if (block.type === "tool_use" && block.name === toolName) {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      proposal = block.input as Record<string, any>;
    }
  }
  return { reply, proposal };
}

const GEMINI_TYPE_MAP: Record<string, Type> = {
  object: Type.OBJECT,
  string: Type.STRING,
  integer: Type.INTEGER,
  number: Type.NUMBER,
  array: Type.ARRAY,
  boolean: Type.BOOLEAN,
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function toGeminiSchema(schema: JsonSchema): Record<string, any> {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const out: Record<string, any> = {};
  if (schema.type) out.type = GEMINI_TYPE_MAP[schema.type] ?? schema.type;
  if (schema.enum) out.enum = schema.enum;
  if (schema.properties) {
    out.properties = Object.fromEntries(
      Object.entries(schema.properties).map(([k, v]) => [k, toGeminiSchema(v as JsonSchema)]),
    );
  }
  if (schema.items) out.items = toGeminiSchema(schema.items);
  if (schema.required) out.required = schema.required;
  return out;
}

async function runGeminiTurn(
  systemPrompt: string,
  history: AgentChatMessage[],
  stateContext: string,
  toolName: string,
  _toolDescription: string,
  inputSchema: JsonSchema,
): Promise<StructuredCallResult> {
  const apiKey = process.env.GEMINI_API_KEY;
  if (!apiKey) {
    throw new AIProviderError("GEMINI_API_KEY is not configured");
  }
  const client = new GoogleGenAI({ apiKey });

  const conversation = [stateContext, ...history.map((m) => `${m.role}: ${m.content}`)].join("\n\n");

  let response;
  try {
    response = await client.models.generateContent({
      model: GEMINI_MODEL,
      contents: conversation,
      config: {
        systemInstruction: `${systemPrompt}\n\nAlways respond with a JSON object having two top-level keys: "reply" (your conversational text response) and "${toolName}" (the complete proposed state if you're proposing a change, or null if you're just answering a question).`,
        responseMimeType: "application/json",
        responseSchema: {
          type: Type.OBJECT,
          properties: {
            reply: { type: Type.STRING },
            [toolName]: toGeminiSchema(inputSchema),
          },
          required: ["reply"],
        },
      },
    });
  } catch (err) {
    throw new AIProviderError(`Gemini API call failed: ${err instanceof Error ? err.message : err}`);
  }

  const text = response.text;
  if (!text) {
    throw new AIProviderError("Gemini model returned no content");
  }

  let parsed: Record<string, unknown>;
  try {
    parsed = JSON.parse(text);
  } catch (err) {
    throw new AIProviderError(`Gemini response was not valid JSON: ${err instanceof Error ? err.message : err}`);
  }

  const reply = typeof parsed.reply === "string" ? parsed.reply : "";
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const proposal = (parsed[toolName] as Record<string, any> | null | undefined) ?? null;
  return { reply, proposal };
}

File: apps/aiwebmaster/core/ai_provider.py
"""
AI provider abstraction — same pattern as apps/api/core/ai_provider.py and
apps/cms/src/lib/aiProvider.ts, kept as its own copy since this is a separate
deployable service with no shared package to import from.

Provider/model/key resolve DB-first (core/ai_settings.py, configurable from
the Settings UI) falling back to env vars (core/config.py) so an env-only
deploy keeps working untouched until someone configures via the UI.
"""
from __future__ import annotations

import json
from typing import Any

from core.ai_settings import get_active_ai_settings
from core.config import settings

DEFAULT_MODELS = {
    "anthropic": "claude-sonnet-4-5",
    "gemini": "gemini-flash-latest",
    "openai": "gpt-4o",
}

_GEMINI_TYPE_MAP = {
    "object": "OBJECT",
    "string": "STRING",
    "integer": "INTEGER",
    "number": "NUMBER",
    "array": "ARRAY",
    "boolean": "BOOLEAN",
}


class AIProviderError(Exception):
    pass


def _active_config() -> dict[str, str]:
    """Resolves {provider, model, api_key} — DB row if configured, else env vars."""
    row = get_active_ai_settings()
    if row:
        return {"provider": row["provider"], "model": row["model"], "api_key": row["api_key"]}

    provider = settings.ai_provider.lower()
    key = settings.anthropic_api_key if provider == "anthropic" else settings.gemini_api_key
    return {"provider": provider, "model": DEFAULT_MODELS.get(provider, ""), "api_key": key}


def list_models(provider: str, api_key: str) -> list[str]:
    """Live model list from the provider's API, falling back to a curated
    static list if the provider has no list-models endpoint reachable here."""
    provider = provider.lower()
    try:
        if provider == "anthropic":
            import anthropic

            client = anthropic.Anthropic(api_key=api_key)
            return [m.id for m in client.models.list(limit=50).data]
        if provider == "openai":
            import openai

            client = openai.OpenAI(api_key=api_key)
            ids = [m.id for m in client.models.list().data]
            return sorted(m for m in ids if "gpt" in m or "o1" in m or "o3" in m) or ids
        if provider == "gemini":
            from google import genai

            client = genai.Client(api_key=api_key)
            return [m.name.replace("models/", "") for m in client.models.list()]
    except Exception as exc:
        raise AIProviderError(f"Could not fetch models for {provider}: {exc}") from exc
    raise AIProviderError(f"Unknown provider '{provider}'")


def _to_gemini_schema(schema: dict[str, Any]) -> dict[str, Any]:
    out: dict[str, Any] = {}
    if "type" in schema:
        out["type"] = _GEMINI_TYPE_MAP.get(schema["type"], str(schema["type"]).upper())
    if "enum" in schema:
        out["enum"] = schema["enum"]
    if "properties" in schema:
        out["properties"] = {k: _to_gemini_schema(v) for k, v in schema["properties"].items()}
    if "items" in schema:
        out["items"] = _to_gemini_schema(schema["items"])
    if "required" in schema:
        out["required"] = schema["required"]
    return out


def structured_call(
    *,
    system_prompt: str,
    user_message: str,
    tool_name: str,
    tool_description: str,
    input_schema: dict[str, Any],
    max_tokens: int = 2048,
) -> dict[str, Any]:
    """Returns {"reply": str, "proposal": dict | None}."""
    cfg = _active_config()
    if cfg["provider"] == "gemini":
        return _call_gemini(cfg, system_prompt, user_message, tool_name, input_schema, max_tokens)
    if cfg["provider"] == "openai":
        return _call_openai(cfg, system_prompt, user_message, tool_name, tool_description, input_schema, max_tokens)
    return _call_anthropic(cfg, system_prompt, user_message, tool_name, tool_description, input_schema, max_tokens)


def _call_anthropic(
    cfg: dict[str, str],
    system_prompt: str,
    user_message: str,
    tool_name: str,
    tool_description: str,
    input_schema: dict[str, Any],
    max_tokens: int,
) -> dict[str, Any]:
    import anthropic

    if not cfg["api_key"]:
        raise AIProviderError("No Anthropic API key configured (Settings page or ANTHROPIC_API_KEY)")

    tool = {"name": tool_name, "description": tool_description, "input_schema": input_schema}
    try:
        client = anthropic.Anthropic(api_key=cfg["api_key"])
        response = client.messages.create(
            model=cfg["model"] or DEFAULT_MODELS["anthropic"],
            max_tokens=max_tokens,
            system=system_prompt,
            tools=[tool],
            messages=[{"role": "user", "content": user_message}],
        )
    except anthropic.APIError as exc:
        raise AIProviderError(f"Anthropic API call failed: {exc}") from exc

    reply = ""
    proposal = None
    for block in response.content:
        if block.type == "text":
            reply += block.text
        elif block.type == "tool_use" and block.name == tool_name:
            proposal = block.input
    return {"reply": reply, "proposal": proposal}


def _call_openai(
    cfg: dict[str, str],
    system_prompt: str,
    user_message: str,
    tool_name: str,
    tool_description: str,
    input_schema: dict[str, Any],
    max_tokens: int,
) -> dict[str, Any]:
    import openai

    if not cfg["api_key"]:
        raise AIProviderError("No OpenAI API key configured (Settings page)")

    tool = {
        "type": "function",
        "function": {"name": tool_name, "description": tool_description, "parameters": input_schema},
    }
    try:
        client = openai.OpenAI(api_key=cfg["api_key"])
        response = client.chat.completions.create(
            model=cfg["model"] or DEFAULT_MODELS["openai"],
            max_tokens=max_tokens,
            messages=[
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": user_message},
            ],
            tools=[tool],
        )
    except openai.APIError as exc:
        raise AIProviderError(f"OpenAI API call failed: {exc}") from exc

    message = response.choices[0].message
    reply = message.content or ""
    proposal = None
    if message.tool_calls:
        for call in message.tool_calls:
            if call.function.name == tool_name:
                try:
                    proposal = json.loads(call.function.arguments)
                except json.JSONDecodeError as exc:
                    raise AIProviderError(f"OpenAI returned invalid tool arguments JSON: {exc}") from exc
                break
    return {"reply": reply, "proposal": proposal}


def _call_gemini(
    cfg: dict[str, str],
    system_prompt: str,
    user_message: str,
    tool_name: str,
    input_schema: dict[str, Any],
    max_tokens: int,
) -> dict[str, Any]:
    from google import genai
    from google.genai import types

    if not cfg["api_key"]:
        raise AIProviderError("No Gemini API key configured (Settings page or GEMINI_API_KEY)")

    response_schema = {
        "type": "OBJECT",
        "properties": {
            "reply": {"type": "STRING"},
            tool_name: _to_gemini_schema(input_schema),
        },
        "required": ["reply"],
    }
    system = (
        f"{system_prompt}\n\nAlways respond with a JSON object having two top-level keys: "
        f'"reply" (your conversational text) and "{tool_name}" (the complete proposed actions '
        "list if you're proposing changes, or null if you're just answering a question)."
    )
    try:
        client = genai.Client(api_key=cfg["api_key"])
        response = client.models.generate_content(
            model=cfg["model"] or DEFAULT_MODELS["gemini"],
            contents=user_message,
            config=types.GenerateContentConfig(
                system_instruction=system,
                max_output_tokens=max_tokens,
                response_mime_type="application/json",
                response_schema=response_schema,
            ),
        )
    except Exception as exc:
        raise AIProviderError(f"Gemini API call failed: {exc}") from exc

    if not response.text:
        raise AIProviderError("Gemini model returned no content")
    try:
        parsed = json.loads(response.text)
    except json.JSONDecodeError as exc:
        raise AIProviderError(f"Gemini response was not valid JSON: {exc}") from exc

    return {"reply": parsed.get("reply", ""), "proposal": parsed.get(tool_name)}
